Cannot build project when special character presents in the path of the project

### Description of the bug / feature
Got the following error when trying to run/build an application whose path has a special character.
```
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:22.0.0.alpha3:prepare-frontend (default) on project special-character: Could not execute prepare-frontend goal.: Cannot invoke "String.replace(char, char)" because "path" is null -> [Help 1]
```
### Minimal reproducible example
1. Create a folder which has special characters, e.g. `äbc`
2. Download a Vaadin 22 Flow starter project (could just use the default one)
3. Put the zip file into the `äbc` folder, unzip and run the application
### Expected behavior
The application runs
### Actual behavior
Got the above mentioned error message and the application cannot start
### Versions:
    - Vaadin / Flow version: Vaadin 22 alpha3
    - Java version: openJDK Runtime Environment Corretto-16.0.1.9.1 (build 16.0.1+9)
    - OS version: macOs big sur 11.5.2

Steps 1-3 give me a bit different errors:

when running with `mvn`:

```
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:2.5.4:run (default-cli) on project special-char: Could not exec java: Cannot run program "/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/bin/java" (in directory "/Users/mikhail/workspace/vaadin/starters/special-chars-?bc/special-char"): error=2, No such file or directory -> [Help 1]
```

when running with `Run 'Application'` in IntellijIDEA:
```
2021-09-21 10:21:47.717 ERROR 2668 --- [  restartedMain] o.a.c.c.C.[Tomcat].[localhost].[/]       : Exception sending context initialized event to listener instance of class [com.vaadin.flow.spring.VaadinServletContextInitializer$CompositeServletContextListener]

java.lang.RuntimeException: Unable to initialize com.vaadin.flow.spring.VaadinServletContextInitializer$VaadinAppShellContextListener
	at com.vaadin.flow.spring.VaadinServletContextInitializer$FailFastServletContextListener.contextInitialized(VaadinServletContextInitializer.java:193) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$CompositeServletContextListener.lambda$contextInitialized$0(VaadinServletContextInitializer.java:210) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) ~[na:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$CompositeServletContextListener.contextInitialized(VaadinServletContextInitializer.java:210) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4766) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5230) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140) ~[na:na]
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:835) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140) ~[na:na]
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:263) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:432) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:927) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.apache.catalina.startup.Tomcat.start(Tomcat.java:486) ~[tomcat-embed-core-9.0.52.jar:9.0.52]
	at org.springframework.boot.web.embedded.tomcat.TomcatWebServer.initialize(TomcatWebServer.java:123) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.web.embedded.tomcat.TomcatWebServer.<init>(TomcatWebServer.java:104) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getTomcatWebServer(TomcatServletWebServerFactory.java:450) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getWebServer(TomcatServletWebServerFactory.java:199) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:182) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:160) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:577) ~[spring-context-5.3.9.jar:5.3.9]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:434) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:338) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) ~[spring-boot-2.5.4.jar:2.5.4]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1332) ~[spring-boot-2.5.4.jar:2.5.4]
	at com.example.application.Application.main(Application.java:27) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:564) ~[na:na]
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49) ~[spring-boot-devtools-2.5.4.jar:2.5.4]
Caused by: java.lang.IllegalStateException: Parent resource of [/Users/mikhail/workspace/vaadin/starters/special-chars-äbc/special-char/target/classes/com/example/application/Application.class] not found in the resources!
	at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.collectResources(VaadinServletContextInitializer.java:845) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.getResources(VaadinServletContextInitializer.java:815) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:420) ~[spring-context-5.3.9.jar:5.3.9]
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.findCandidateComponents(ClassPathScanningCandidateComponentProvider.java:315) ~[spring-context-5.3.9.jar:5.3.9]
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[na:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$VaadinAppShellContextListener.failFastContextInitialized(VaadinServletContextInitializer.java:553) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	at com.vaadin.flow.spring.VaadinServletContextInitializer$FailFastServletContextListener.contextInitialized(VaadinServletContextInitializer.java:189) ~[vaadin-spring-19.0.0.alpha1.jar:na]
	... 46 common frames omitted
```

I can still reproduce this class of path issue with current Vaadin 25.x.

The failure is reproducible on macOS when the project path contains a decomposed Unicode character, specifically `c` plus combining cedilla (`U+0063 U+0327`). The same minimal project passes from an ASCII path and from a path containing only a regular space.

This looks related to the older reports in #8191 and #18445, but those are closed. I am adding this here instead of opening a duplicate because this issue is still open and an earlier comment already shows the same `Parent resource of [...] not found in the resources!` failure class.

### Reproduction

Create a minimal Vaadin Spring Boot project:

```xml
<!-- pom.xml -->
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>4.0.6</version>
        <relativePath/>
    </parent>

    <groupId>com.example</groupId>
    <artifactId>vaadin-path-repro</artifactId>
    <version>1.0.0-SNAPSHOT</version>

    <properties>
        <java.version>25</java.version>
        <vaadin.version>25.1.4</vaadin.version>
    </properties>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.vaadin</groupId>
                <artifactId>vaadin-bom</artifactId>
                <version>${vaadin.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>com.vaadin</groupId>
            <artifactId>vaadin-spring-boot-starter</artifactId>
        </dependency>
        <dependency>
            <groupId>com.vaadin</groupId>
            <artifactId>vaadin-dev</artifactId>
            <optional>true</optional>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>com.vaadin</groupId>
                <artifactId>vaadin-maven-plugin</artifactId>
                <version>${vaadin.version}</version>
            </plugin>
        </plugins>
    </build>
</project>
```

```java
// src/main/java/com/example/application/Application.java
package com.example.application;

import com.vaadin.flow.component.page.AppShellConfigurator;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class Application implements AppShellConfigurator {
    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }
}
```

```java
// src/main/java/com/example/application/MainView.java
package com.example.application;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.router.Route;

@Route("")
public class MainView extends Div {
    public MainView() {
        setText("Hello");
    }
}
```

```java
// src/test/java/com/example/application/ApplicationTest.java
package com.example.application;

import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ApplicationTest {
    @Test
    void contextLoads() {
    }
}
```

Then run the same project from different paths:

```bash
# Baseline ASCII path: passes
cd /tmp/vaadin-path-repro
mvn -q test

# Path containing only a regular space: also passes
cp -R /tmp/vaadin-path-repro "/tmp/vaadin space only"
cd "/tmp/vaadin space only"
mvn -q test

# Path containing decomposed Unicode: fails
python3 - <<'PY'
import shutil
from pathlib import Path

src = Path("/tmp/vaadin-path-repro")
dst = Path("/tmp") / "vaadin-c\u0327-repro"
shutil.rmtree(dst, ignore_errors=True)
shutil.copytree(src, dst)
print(dst)
PY

cd "$(python3 - <<'PY'
from pathlib import Path
print(Path("/tmp") / "vaadin-c\u0327-repro")
PY
)"
mvn -q test
```

### Actual result

The decomposed Unicode path fails during Spring Boot test context startup:

```text
java.lang.RuntimeException: Unable to initialize com.vaadin.flow.spring.VaadinServletContextInitializer$VaadinAppShellContextListener
...
Caused by: java.lang.IllegalStateException: Parent resource of [/private/tmp/vaadin-c%CC%A7-repro/target/test-classes/com/example/application/ApplicationTest.class] not found in the resources!
    at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.collectResources(VaadinServletContextInitializer.java:1098) ~[vaadin-spring-25.1.4.jar:na]
    at com.vaadin.flow.spring.VaadinServletContextInitializer$CustomResourceLoader.getResources(VaadinServletContextInitializer.java:1034) ~[vaadin-spring-25.1.4.jar:na]
```

The encoded `%CC%A7` is the combining cedilla in the path.

### Expected result

The application should start from the decomposed Unicode path in the same way it starts from the ASCII path and the path containing only a regular space.

### Versions

- Vaadin / Flow version: reproduced with `25.1.3` and `25.1.4`
- Java version: Azul Zulu `25.0.1`
- Maven version: `3.9.11`
- OS version: macOS `26.3.1`, Apple Silicon
- Browser version: n/a
- Application server: embedded Tomcat `11.0.21` via Spring Boot `4.0.6`

### Related issues

- #8191 reported a similar path encoding problem and was closed as not reproducible in Vaadin 23.3.
- #18445 reported a space-in-path production JAR issue and was closed after maintainers could not reproduce it.
- #18730 had the same error message but was tied to dev-mode reload/cache behavior and was fixed in #18754.

AI assistance disclosure: I used an AI assistant to help reproduce the issue, compare it against existing reports, and draft this comment. I reviewed the reproduction result and content before submitting.

--- a/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
+++ b/vaadin-spring/src/main/java/com/vaadin/flow/spring/VaadinServletContextInitializer.java
@@ -26,6 +26,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.Serializable;
 import java.lang.annotation.Annotation;
+import java.net.URISyntaxException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
@@ -976,8 +977,7 @@ public class VaadinServletContextInitializer
      * For npm we scan all packages. For performance reasons and due to problems
      * with atmosphere we skip known packaged from our resources collection.
      */
-    private static class CustomResourceLoader
-            extends FilterableResourceResolver {
+    static class CustomResourceLoader extends FilterableResourceResolver {
 
         private final PrefixTree scanNever = new PrefixTree(DEFAULT_SCAN_NEVER);
 
@@ -1048,15 +1048,8 @@ public class VaadinServletContextInitializer
 
             for (Resource resource : super.getResources(locationPattern)) {
                 String originalPath = resource.getURL().getPath();
-                String path;
-                if (originalPath.startsWith("file:///resources!")) {
-                    // It's a resource from a native build, remove the
-                    // prefix from URL path
-                    path = originalPath
-                            .substring("file:///resources!".length());
-                } else {
-                    path = originalPath;
-                }
+                final String path = getComparableResourcePath(resource,
+                        originalPath);
 
                 if (isDevModeCacheUsed() && skipped.contains(originalPath)) {
                     continue;
@@ -1066,8 +1059,8 @@ public class VaadinServletContextInitializer
                     resources.add(resource);
                     // Restore root paths to ensure new resources are correctly
                     // validate and cached after a reload
-                    if (originalPath.endsWith("/")) {
-                        rootPaths.add(originalPath);
+                    if (path.endsWith("/")) {
+                        rootPaths.add(path);
                     }
                 } else {
                     if (path.endsWith(".jar!/")) {
@@ -1122,6 +1115,26 @@ public class VaadinServletContextInitializer
             }
 
             return resources.toArray(new Resource[0]);
+        }
+
+        private String getComparableResourcePath(Resource resource,
+                String fallbackPath) throws IOException {
+            try {
+                String path = resource.getURL().toURI().getPath();
+                return stripNativeImageResourcePrefix(
+                        path == null ? fallbackPath : path);
+            } catch (IllegalArgumentException | URISyntaxException exception) {
+                return stripNativeImageResourcePrefix(fallbackPath);
+            }
+        }
+
+        private String stripNativeImageResourcePrefix(String path) {
+            if (path.startsWith("file:///resources!")) {
+                // It's a resource from a native build, remove the prefix from
+                // URL path.
+                return path.substring("file:///resources!".length());
+            }
+            return path;
         }
 
         private boolean isDevModeCacheUsed() {

--- a/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
+++ b/vaadin-spring/src/test/java/com/vaadin/flow/spring/VaadinServletContextInitializerTest.java
@@ -19,6 +19,12 @@ import jakarta.servlet.ServletContext;
 import jakarta.servlet.ServletContextEvent;
 import jakarta.servlet.ServletContextListener;
 
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
@@ -29,6 +35,7 @@ import java.util.stream.Stream;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 import org.mockito.Mock;
 import org.mockito.MockedStatic;
 import org.mockito.Mockito;
@@ -36,6 +43,8 @@ import org.mockito.MockitoAnnotations;
 import org.springframework.boot.autoconfigure.AutoConfigurationPackages;
 import org.springframework.context.ApplicationContext;
 import org.springframework.core.env.Environment;
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.Resource;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.di.Lookup;
@@ -53,6 +62,7 @@ import com.vaadin.flow.server.startup.ApplicationConfiguration;
 import com.vaadin.flow.server.startup.ApplicationRouteRegistry;
 import com.vaadin.flow.server.startup.ServletDeployer;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 class VaadinServletContextInitializerTest {
@@ -214,6 +224,36 @@ class VaadinServletContextInitializerTest {
                 .getErrorNavigationTarget(new NotFoundException()).get()
                 .getNavigationTarget();
         assertEquals(TestErrorView.class, navigationTarget);
+    }
+
+    @Test
+    void customResourceLoader_classpathRootContainsUnicodeCombiningCharacter_resourcesAreMatched(
+            @TempDir Path tempDir) throws Exception {
+        Path classesRoot = tempDir.resolve("François")
+                .resolve("vaadin-c\u0327-repro").resolve("target/classes");
+        Path applicationClass = classesRoot
+                .resolve("com/example/application/Application.class");
+        Files.createDirectories(applicationClass.getParent());
+        Files.write(applicationClass, new byte[] { 0 });
+
+        try (URLClassLoader classLoader = new URLClassLoader(
+                new URL[] { classesRoot.toUri().toURL() }, null)) {
+            Resource[] resources = new VaadinServletContextInitializer.CustomResourceLoader(
+                    new DefaultResourceLoader(classLoader)).getResources(
+                            "classpath*:com/example/application/**/*.class");
+
+            assertThat(resources).as(Arrays.toString(resources)).anySatisfy(
+                    resource -> assertThat(getResourcePath(resource)).endsWith(
+                            "/com/example/application/Application.class"));
+        }
+    }
+
+    private static String getResourcePath(Resource resource) {
+        try {
+            return resource.getURI().getPath();
+        } catch (IOException exception) {
+            throw new IllegalStateException(exception);
+        }
     }
 
     private Runnable initRouteNotFoundMocksAndGetContextInitializedMockCall(

